### PR TITLE
How to implement Data abstraction in C++

### LINKED ISSUE #1778: Understanding abstraction in c++
Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

### Proposed title of the article
Understanding abstraction in c++

### Introduction paragraph (2-3 paragraphs):
Data abstraction is one of the most essential and important features of object-oriented programming. Data abstraction concept in c++ helps developers to provide only essential information without providing the background details. It is the most widely used technique that relies on the separation of implementation and interface of the code. This article will take a look at data abstraction along with the types and respective examples.

### Key takeaways: 
1. What is abstraction
2. Types of abstraction
3. Advantages of data abstraction
4. How to achieve data abstraction in c++ through class, header files, and access specifiers

### References: 
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### LINKED ISSUE #2060: How to implement Data abstraction in C++
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
How to implement Data abstraction in C++

### Introduction paragraph (2-3 paragraphs):
Abstraction is the act of showing essential features to the user but hiding the background details. Data abstraction refers to providing only essential information about the data to the outside world. In C++, classes provide a great level of data abstraction. They provide enough methods to the outside world to play with the functionality of the object.  And to manipulate object data state without actually knowing how the class is implemented. For example, we can use the `pow()` function to calculate the power of a number without knowing the algorithm the function follows. In fact, the underlying implementation of the algorithm could change between releases of the library. As long as the interface stays the same, your function call will still work. This article will go through an introduction to abstraction along with types and how data abstraction is achieved in C++.

### Key takeaways: 
1. Types Of Abstraction
2. Abstraction Using Specifiers
3. Abstraction using class's
4. Abstraction using header files
5. Advantages of data abstraction

### References: 
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Dawe-7**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #2060

### CODE REVIEW

**Review by kayere — APPROVED** [review]
Great article @Dawe-7.
@hectorkambow - Please give this article a final review.

File: articles/understanding-abstraction-in-c++/index.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+---
+### Introduction
+Abstraction is the act of representing essential features without including the background details. `Data abstraction` refers to providing only essential information about the data to the outside world while hiding the background details or explanations. In C++, classes provide a great level of data abstraction. They provide enough public methods to the outside world to play with the functionality of the object and to manipulate object data.
+<!--more-->
+For example, a state without actually knowing how the class has been implemented internally. For example, the `pow()` function is used to calculate the power of a number without knowing the algorithm the function follows. In fact, the underlying implementation of the sorting functionality could change between releases of the library. And as long as the interface stays the same, your function call will still work. This article will go through an introduction to abstraction along with types and how data abstraction is achieved in C++.
+
+### Prerequisites
+To follow this article, you’ll need to have:
+
+- [Codeblocks IDE]() to run the code.
```
Comment (M0nicah) [inline articles/understanding-abstraction-in-c++/index.md:25]: Almost forgot, please add this link.

Comment (kayere) [inline articles/understanding-abstraction-in-c++/index.md:25]: Almost forgot. Please add this link.

Comment (Dawe-7) [reply articles/understanding-abstraction-in-c++/index.md]: Hi, @kayere I've added the link, thanks for the review.

**Comment by hectorkambow** [comment]
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.

[How to implement Data abstraction in C++ #2148.pdf](https://github.com/section-engineering-education/engineering-education/files/6443570/How.to.implement.Data.abstraction.in.C%2B%2B.2148.pdf)

**Comment by Dawe-7** [comment]
> Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
> Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
> Be sure to see our resources page to see more info on plagiarism and what is considered as such.
> 
> [How to implement Data abstraction in C++ #2148.pdf](https://github.com/section-engineering-education/engineering-education/files/6443570/How.to.implement.Data.abstraction.in.C%2B%2B.2148.pdf)

Hello, @hectorkambow I've fixed the plagiarism. Kindly review.

cc: @kayere

**Comment by hectorkambow** [comment]
Thank you v much for your patience as we worked through our Repo re-structuring process. 

The purpose behind this restructure - was to better allow us to handle more volume and allow students to publish more articles and get more involved with the community (with open-source contribution and suggestions). 

With most of the changes coming to completion - we will be sure to get your article published asap. 

Thank you again - and we are looking forward to more amazing articles.

**Comment by hectorkambow** [comment]
This line was empty
>Go ahead run the code [here]()

@Dawe-7 

cc @kayere 

I have also migrated the files to a new PR to prevent a merge conflict with the new folder structure

**Pull request closed without merging**